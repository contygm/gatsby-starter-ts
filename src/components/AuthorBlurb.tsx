import * as React from 'react';
import { StaticImage } from 'gatsby-plugin-image';
import FaIconLink from '../components/FaIconLink';
import { faArrowLeft, faArrowRight } from '@fortawesome/free-solid-svg-icons';

export interface AuthorBlurbProps {
    author: {
        name: string;
        description: string;
    };
    postTags: Array<string>;
    postDate: string;
    previousPost: NeighborPost;
    nextPost: NeighborPost;
}

const AuthorMedia = ({ name, description }: AuthorBlurbProps['author']) => {
    return (
        <div className="media p-2 m-2">
            <figure className="media-left ml-0 mb-0">
                <StaticImage
                    src="../images/icon.png"
                    alt="author image"
                    placeholder="blurred"
                    className="image is-96x96"
                />
            </figure>
            <div className="media-content m-auto">
                <div className="content">
                    <p>
                        <strong>{name}</strong>
                        <br />
                        {description}
                    </p>
                </div>
            </div>
        </div>
    );
};

const DateAndTags = (props: { tags: Array<string>; date: string }) => {
    return (
        <div className="level mb-2">
            <div className="level-left">
                <div className="level-item">
                    <p>
                        Published: <time>{props.date}</time>
                    </p>
                </div>
            </div>
            <div className="level-right">
                <div className="level-item">
                    {props.tags.map((tag) => {
                        return (
                            <a
                                href="/404"
                                className="post-tag"
                                key={tag}
                            >
                                {tag}
                            </a>
                        );
                    })}
                </div>
            </div>
        </div>
    );
};

const NextAndPrevious = (props: {
    next: NeighborPost;
    previous: NeighborPost;
}) => {
    return (
        <div className="level my-3">
            <div className="level-left">
                <div className="level-item">
                    {props.previous && (
                        <FaIconLink
                            faIcon={faArrowLeft}
                            url={`/blog${props.previous.fields.slug}`}
                            label={props.previous.frontmatter.title}
                            labelLeft={false}
                        />
                    )}
                </div>
            </div>
            <div className="level-right">
                <div className="level-item">
                    {props.next && (
                        <FaIconLink
                            faIcon={faArrowRight}
                            url={`/blog${props.next.fields.slug}`}
                            label={props.next.frontmatter.title}
                        />
                    )}
                </div>
            </div>
        </div>
    );
};

const AuthorBlurb = (props: AuthorBlurbProps) => {
    return (
        <section className="section">
            <div className="container author-container">
                <DateAndTags
                    tags={props.postTags}
                    date={props.postDate}
                />
                <AuthorMedia
                    name={props.author.name}
                    description={props.author.description}
                />
                <NextAndPrevious
                    next={props.nextPost}
                    previous={props.previousPost}
                />
            </div>
        </section>
    );
};

export default AuthorBlurb;
