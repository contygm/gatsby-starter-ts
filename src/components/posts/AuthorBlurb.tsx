import * as React from 'react';
import { StaticImage } from 'gatsby-plugin-image';
import { Link } from 'gatsby';
import { NextAndPrevious } from '../common/NextAndPrevious';

/**
 * All the properties needed for the Author section at the end of a post
 * @interface AuthorBlurbProps
 *
 * @property {Object} author - An object containing the author name and description
 * @property {string} author.name - the name of the author
 * @property {string} author.description - the description of the author
 * @property {string[]} postTags - A list of all the tags for the post
 * @property {string} postDate - the post's date
 * @property {NeighborPost} [previousPost] - the post published immediately *before* this post (sorted by postDate)
 * @property {NeighborPost} [nextPost] - the post published immediately *after* this post (sorted by postDate)
 *
 * @see AuthorBlurb
 */
// TODO group objects for easy typing on sub components
export interface AuthorBlurbProps {
    author: {
        name: string;
        description: string;
    };
    postTags: string[];
    postDate: string;
    previousPost?: NeighborPost;
    nextPost?: NeighborPost;
}

/**
 * The author media sub-component uses a placeholder author image and displays the author name + description next to the image.
 * @memberof AuthorBlurb
 *
 * @param {AuthorBlurbProps.author} - the author's name and description
 */
const AuthorMedia = ({ name, description }: AuthorBlurbProps['author']) => {
    return (
        <div className="media p-2 m-2">
            <figure className="media-left ml-0 mb-0">
                <StaticImage
                    src="../../images/icon.png"
                    alt="author image"
                    placeholder="blurred"
                    className="image is-96x96"
                />
            </figure>
            <div className="media-content m-auto">
                <div className="content">
                    <p>
                        <strong>{name}</strong>
                        <br />
                        {description}
                    </p>
                </div>
            </div>
        </div>
    );
};

/**
 * A row with the post's publish date and all the tags. This is the first row in the author blurb component
 * @memberof AuthorBlurb
 * @param { AuthorBlurbProps.postTags } postTags - all post tags
 * @param { AuthorBlurbProps.postDate } postDate - the post's publication date
 */
const DateAndTags = ({
    postTags,
    postDate
}: {
    postTags: string[];
    postDate: string;
}) => {
    return (
        <div className="level mb-2">
            <div className="level-left">
                <div className="level-item">
                    <p>
                        Published: <time>{postDate}</time>
                    </p>
                </div>
            </div>
            <div className="level-right">
                <div className="level-item">
                    {postTags.map((tag) => {
                        return (
                            <Link
                                to={`/blog?tag=${tag}`}
                                className="post-tag"
                                key={tag}
                            >
                                {tag}
                            </Link>
                        );
                    })}
                </div>
            </div>
        </div>
    );
};

/**
 * A component to display author info and some post info at the bottom of the post. The post's tags and publish date
 * are displayed on the first row. Then the author image + description are on the second row. The last row contains
 * links to the next and previous post (sorted by date).
 * @category Components
 * @param {AuthorBlurbProps} props - includes the author info as well as post tags + date. Also includes next and previous post
 *
 * @see AuthorBlurbProps
 */
export const AuthorBlurb = (props: AuthorBlurbProps) => {
    return (
        <section className="section">
            <div className="container author-container">
                <DateAndTags
                    postTags={props.postTags}
                    postDate={props.postDate}
                />
                <AuthorMedia
                    name={props.author.name}
                    description={props.author.description}
                />
                <NextAndPrevious
                    next={props.nextPost}
                    previous={props.previousPost}
                    type={'blog'}
                />
            </div>
        </section>
    );
};
