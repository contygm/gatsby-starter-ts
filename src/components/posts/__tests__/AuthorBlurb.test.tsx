import React from 'react';
import { render } from '@testing-library/react';
import { AuthorBlurb, AuthorBlurbProps } from '../AuthorBlurb';

const mockData: AuthorBlurbProps = {
    author: {
        name: 'Luke Skywalker',
        description: "Let's save the galaxy one more time!"
    },
    postTags: ['one', 'two', 'three'],
    postDate: 'March 28, 2000',
    previousPost: {
        fields: {
            slug: '/blog-post-before'
        },
        frontmatter: {
            title: 'Before - A Blog About Things'
        }
    },
    nextPost: {
        fields: {
            slug: '/blog-post-before'
        },
        frontmatter: {
            title: 'Next - A Blog About Things'
        }
    }
};

describe('PageHeader', () => {
    let testProps: AuthorBlurbProps;

    beforeEach(() => {
        testProps = mockData;
    });

    it('default renders correctly', () => {
        const { asFragment } = render(<AuthorBlurb {...testProps} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('renders when previousPost is undefined', () => {
        testProps.previousPost = undefined;
        const { asFragment } = render(<AuthorBlurb {...testProps} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('renders when previousPost is defined', () => {
        testProps.nextPost = undefined;
        const { asFragment } = render(<AuthorBlurb {...testProps} />);
        expect(asFragment()).toMatchSnapshot();
    });
});
