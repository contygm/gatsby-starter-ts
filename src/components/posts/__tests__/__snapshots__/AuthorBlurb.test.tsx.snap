// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageHeader default renders correctly 1`] = `
<DocumentFragment>
  <section
    class="section"
  >
    <div
      class="container author-container"
    >
      <div
        class="level mb-2"
      >
        <div
          class="level-left"
        >
          <div
            class="level-item"
          >
            <p>
              Published: 
              <time>
                March 28, 2000
              </time>
            </p>
          </div>
        </div>
        <div
          class="level-right"
        >
          <div
            class="level-item"
          >
            <a
              class="post-tag"
              href="/blog?tag=one"
            >
              one
            </a>
            <a
              class="post-tag"
              href="/blog?tag=two"
            >
              two
            </a>
            <a
              class="post-tag"
              href="/blog?tag=three"
            >
              three
            </a>
          </div>
        </div>
      </div>
      <div
        class="media p-2 m-2"
      >
        <figure
          class="media-left ml-0 mb-0"
        >
          <img
            alt="author image"
            placeholder="blurred"
            src="../../images/icon.png"
          />
        </figure>
        <div
          class="media-content m-auto"
        >
          <div
            class="content"
          >
            <p>
              <strong>
                Luke Skywalker
              </strong>
              <br />
              Let's save the galaxy one more time!
            </p>
          </div>
        </div>
      </div>
      <div
        class="level my-3"
      >
        <div
          class="level-left"
        >
          <div
            class="level-item"
          >
            <li
              class="faicon-link-container"
            >
              <a
                aria-label="social media link"
                class="faicon-link with-label is-black"
                data-cy="faicon-link"
                href="/blog/blog-post-before"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-arrow-left fa-xl icon"
                  data-icon="arrow-left"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.2 288 416 288c17.7 0 32-14.3 32-32s-14.3-32-32-32l-306.7 0L214.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-160 160z"
                    fill="currentColor"
                  />
                </svg>
                <span>
                  Before - A Blog About Things
                </span>
              </a>
            </li>
          </div>
        </div>
        <div
          class="level-right"
        >
          <div
            class="level-item"
          >
            <li
              class="faicon-link-container"
            >
              <a
                aria-label="social media link"
                class="faicon-link with-label is-black"
                data-cy="faicon-link"
                href="/blog/blog-post-before"
              >
                <span>
                  Next - A Blog About Things
                </span>
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-arrow-right fa-xl icon"
                  data-icon="arrow-right"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M438.6 278.6c12.5-12.5 12.5-32.8 0-45.3l-160-160c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L338.8 224 32 224c-17.7 0-32 14.3-32 32s14.3 32 32 32l306.7 0L233.4 393.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0l160-160z"
                    fill="currentColor"
                  />
                </svg>
              </a>
            </li>
          </div>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;

exports[`PageHeader renders when previousPost is defined 1`] = `
<DocumentFragment>
  <section
    class="section"
  >
    <div
      class="container author-container"
    >
      <div
        class="level mb-2"
      >
        <div
          class="level-left"
        >
          <div
            class="level-item"
          >
            <p>
              Published: 
              <time>
                March 28, 2000
              </time>
            </p>
          </div>
        </div>
        <div
          class="level-right"
        >
          <div
            class="level-item"
          >
            <a
              class="post-tag"
              href="/blog?tag=one"
            >
              one
            </a>
            <a
              class="post-tag"
              href="/blog?tag=two"
            >
              two
            </a>
            <a
              class="post-tag"
              href="/blog?tag=three"
            >
              three
            </a>
          </div>
        </div>
      </div>
      <div
        class="media p-2 m-2"
      >
        <figure
          class="media-left ml-0 mb-0"
        >
          <img
            alt="author image"
            placeholder="blurred"
            src="../../images/icon.png"
          />
        </figure>
        <div
          class="media-content m-auto"
        >
          <div
            class="content"
          >
            <p>
              <strong>
                Luke Skywalker
              </strong>
              <br />
              Let's save the galaxy one more time!
            </p>
          </div>
        </div>
      </div>
      <div
        class="level my-3"
      >
        <div
          class="level-left"
        >
          <div
            class="level-item"
          />
        </div>
        <div
          class="level-right"
        >
          <div
            class="level-item"
          />
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;

exports[`PageHeader renders when previousPost is undefined 1`] = `
<DocumentFragment>
  <section
    class="section"
  >
    <div
      class="container author-container"
    >
      <div
        class="level mb-2"
      >
        <div
          class="level-left"
        >
          <div
            class="level-item"
          >
            <p>
              Published: 
              <time>
                March 28, 2000
              </time>
            </p>
          </div>
        </div>
        <div
          class="level-right"
        >
          <div
            class="level-item"
          >
            <a
              class="post-tag"
              href="/blog?tag=one"
            >
              one
            </a>
            <a
              class="post-tag"
              href="/blog?tag=two"
            >
              two
            </a>
            <a
              class="post-tag"
              href="/blog?tag=three"
            >
              three
            </a>
          </div>
        </div>
      </div>
      <div
        class="media p-2 m-2"
      >
        <figure
          class="media-left ml-0 mb-0"
        >
          <img
            alt="author image"
            placeholder="blurred"
            src="../../images/icon.png"
          />
        </figure>
        <div
          class="media-content m-auto"
        >
          <div
            class="content"
          >
            <p>
              <strong>
                Luke Skywalker
              </strong>
              <br />
              Let's save the galaxy one more time!
            </p>
          </div>
        </div>
      </div>
      <div
        class="level my-3"
      >
        <div
          class="level-left"
        >
          <div
            class="level-item"
          />
        </div>
        <div
          class="level-right"
        >
          <div
            class="level-item"
          >
            <li
              class="faicon-link-container"
            >
              <a
                aria-label="social media link"
                class="faicon-link with-label is-black"
                data-cy="faicon-link"
                href="/blog/blog-post-before"
              >
                <span>
                  Next - A Blog About Things
                </span>
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-arrow-right fa-xl icon"
                  data-icon="arrow-right"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M438.6 278.6c12.5-12.5 12.5-32.8 0-45.3l-160-160c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L338.8 224 32 224c-17.7 0-32 14.3-32 32s14.3 32 32 32l306.7 0L233.4 393.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0l160-160z"
                    fill="currentColor"
                  />
                </svg>
              </a>
            </li>
          </div>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
